### DISCUSSION TITLE
Argument of type "list[A]" cannot be assigned to parameter "x" of type "list[A']" in function "xyz"

### DISCUSSION THREAD

**Discussion by sylann** [microsoft/pyright#5835] [Q&A] [answered]
```py
import typing


class A(typing.TypedDict):
    foo: str
    bar: int
    a_specific: typing.Any


class B(typing.TypedDict):
    foo: str
    bar: int
    b_specific: typing.Any


class TData(typing.TypedDict):  # Common "interface"
    foo: str
    bar: int


def handle_datum(datum: TData):
    print(datum["foo"])
    print(datum["bar"])


def handle_data(data: list[TData]):
    for datum in data:
        print(datum["foo"])
        print(datum["bar"])


dict_a: A = {"foo": "foo", "bar": 42, "a_specific": []}
dict_b: B = {"foo": "foo", "bar": 42, "b_specific": {}}

handle_datum(dict_a)  # ok
handle_datum(dict_b)  # ok

handle_data([dict_a])  # ok
handle_data([dict_b])  # ok

dicts_a = [dict_a]
dicts_b = [dict_b]
handle_data(dicts_a)  # Error (see below)
handle_data(dicts_b)  # Error (see below)
```

```
Argument of type "list[A]" cannot be assigned to parameter "data" of type "list[TData]" in function "handle_data"
  "list[A]" is incompatible with "list[TData]"
    Type parameter "_T@list" is invariant, but "A" is not the same as "TData"
```

Given that `TData` is clearly a valid subtype of both `A` and `B`, how can I handle this type issue?

Theoretically, I could do this for example:
```py
TDATA = typing.Union[A, B]
```

But `TData` is supposed to be a generic requirement in a generic module.
It is definitely not ok to have to declare all use cases's types upfront.

**Comment by erictraut** [comment, marked as answer]
You're correct that `TData` is a subtype of `A` and `B`, but `list[TData]` is not a subtype of `list[A]` or `list[B]` because the type parameter for `list` is invariant. For more details about what this means, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts?id=generic-types).

One straightforward way to avoid this type violation is to change `handle_data` to accept a `Sequence[TData]` rather than a `list[TData]`. By making this change, you're guaranteeing that `handle_data` won't mutate the contents of the object that you're passing it.

```python
def handle_data(data: typing.Sequence[TData]): ...
```

**Reply by sylann** [reply]
Hi @erictraut,

Thanks for your answer!
It makes total sense because the other use cases should not mutate the data indeed.

---

In the mean time, I found a hack using an intermediate generic TypeVar.
I'm not sure if it's a bug or intentional:

```py
...

_TData = typing.TypeVar("_TData", bound=TData)

def handle_data(data: list[_TData]):
    ...

...

handle_data(dicts_a)  # ok
```

**Reply by erictraut** [reply]
That is intentional. You'll see that if `handle_data` attempts to add any new items to the list, they must also be of type `_TData`. So the list is mutable in this case, but it cannot be mutated in a way that results in a type violation. I agree that this is a bit of a hack, as you're using a type variable in a way that it's not really designed to be used, but it is type safe.
